tbot_contrib/locking.py

```diff
@@ -0,0 +1,302 @@
+        """
+        return self.name
+
+    def _make_temp(self, labhost) -> linux.Path:
+        retval, result = labhost.exec("mktemp", "lock.XXX")
+        if retval == 0:
+            tempfile = linux.Path(labhost, result.strip())
+            if self.lock_expiry < self.lock_sleep:  # if lock does not expire
+                labhost.exec("echo", "0", linux.RedirStdout(tempfile))
+            else:  # else write the Unix time timestamp of expiry
+                labhost.exec(
+                    "echo",
+                    str(int(time.time()) + self.lock_expiry),
+                    linux.RedirStdout(tempfile),
+                )
+            # Append the PID of this context's shell
+            labhost.exec("echo", linux.Raw("$$"), linux.Raw(">>"), tempfile)
+            return tempfile
+        else:
+            raise Exception("Could not create tempfile")
+
+    def _locker(self, labhost, lockfile) -> bool:
+        tempfile = self._make_temp(labhost)
+        success = False
+        repeat = True
+        while repeat:  # do-while
+            repeat = False
+            print(f"Try locking on {lockfile}...")
+            if labhost.test("ln", tempfile, lockfile):  # hard-linking is atomic
+                success = True
+                repeat = False  # We got a lock, we're done
+            else:  # No lock this time
+                # Read the lock, perhaps it is stale... This may race!
+                retval, result = labhost.exec("cat", lockfile)
+                if retval == 0:
+                    results = result.split()
+                    # Check if the lock can expire, and whether it has expired.
+                    if (int(results[0]) != 0) and (int(results[0]) < int(time.time())):
+                        # Check whether the lock PID is still in use by a bash instance
+                        retval, result = labhost.exec("ps", "-ocomm=", results[1])
+                        if (retval == 0) and (result.strip() == "bash"):
+                            print(f"Expired lock {lockfile} PID still appears valid!")
+                        else:
+                            labhost.exec("rm", lockfile)  # Delete stale lock
+                            print(f"Lock {lockfile} expired, deleted.")
+                            repeat = (
+                                True  # The stale lock may be available next time round.
+                            )
+            if (
+                success
+            ):  # Because things could race, test whether we still have the lock...
+                time.sleep(self.lock_sleep)  # Allow things to settle
+                # Check whether the lockfile still exists and get its inode.
+                retval, result = labhost.exec("stat", "-c%i", lockfile)
+                if retval == 0:  # lock still exists
+                    # Get inode of the tempfile, to see if it matches lockfile's
+                    if result != labhost.exec("stat", "-c%i", tempfile)[1]:
```

Comment:
I think it would be much more readable to use the following construction:

```python
try:
    lockstat = lockfile.stat()
except OsError:
    repeat = True
else:
    if lockstat.st_ino != tempfile.stat().st_ino:
        repeat = True 
```

Reply:
Thanks for that suggestion, I did not know of this handy utility.